README.md
# cool-bayesian-networks

CHANGELOG.md
## v0.0.1 (02/14/2023)

- First release of `cool-bayesian-networks`!


test.py
# Laboratorio 2
# Inteligencia Artificial
# Authors: YongBum Park 20117, Santiago Taracena 20017, Pedro Arriola 20188, Oscar López 20679

def test_bayesian_network_construction():
    assert (2 + 2) == 4, "Test failed"

test_bayesian_network_construction()


cool-bayesian-networks/test.py
# Laboratorio 2
# Inteligencia Artificial
# Authors: YongBum Park 20117, Santiago Taracena 20017, Pedro Arriola 20188, Oscar López 20679

from bayesian_construct import Bayesian

bayesian = Bayesian()

# Network construction
network_structure = [("A", "C"), ("B", "C")]
bayesian.bayesian_network_construction(network_structure)
cpd_a = bayesian.asign_values("A", 2, [[0.3], [0.7]])
cpd_b = bayesian.asign_values("B", 2, [[0.23], [0.77]])
cpd_c = bayesian.asign_values(
    "C", 2, [[0.20, 0.77, 0.10, 0.5], [0.80, 0.23, 0.90, 0.5]], ["A", "B"], [2, 2]
)

def test_bayesian_network_construction():
    assert bayesian.model != None, "The model is not created"

test_bayesian_network_construction()


cool-bayesian-networks/test_bayesian.py
from bayesianConstruct import *

cool-bayesian-networks/utils.py
from bayesian_node import BayesianNode

def replace_all(text, chars, replacement):
    for char in chars:
        text = text.replace(char, replacement)
    return text

def get_probability_dict(line):
    line = [replace_all(probability, [" ", "P", "(", ")", "\n"], "") for probability in line]
    line = "".join([char for char in line if (char != "")])
    line = line.split("=")
    line = [*line[0].split("|"), line[1]]
    return { "probability_of": line[0], "given": line[1], "equals": line[2] } if (len(line) == 3) else { "probability_of": line[0], "given": "", "equals": line[1] }

def get_parent_node(node, probabilities, probability_values, children):

    for probability in probabilities:

        if (node in probability["probability_of"]):
            probability_index = 0 if (probability["probability_of"].startswith("!")) else 1
            probability_value = float(probability["equals"])
            probability_values[0][probability_index] = round(probability_value, 3)
            probability_values[0][int(not probability_index)] = round((1 - probability_value), 3)

    return BayesianNode(
        label=node,
        probability_values=probability_values,
        parents=[],
        children=children,
    )

def get_child_node(node, probabilities, probability_values, parents):
    for probability in probabilities:
        for index, parent in enumerate(parents):
            if (probability["probability_of"] == node and probability["given"] == parent):
                probability_value = float(probability["equals"])
                probability_values[index - 1][1] = round(probability_value, 3)
                probability_values[index - 1][0] = round(1 - probability_value, 3)
            elif (probability["probability_of"] == f"!{node}" and probability["given"] == parent):
                probability_value = float(probability["equals"])
                probability_values[index - 1][0] = round(probability_value, 3)
                probability_values[index - 1][1] = round(1 - probability_value, 3)
            elif (probability["probability_of"] == node and probability["given"] == f"!{parent}"):
                probability_value = float(probability["equals"])
                probability_values[index][1] = round(probability_value, 3)
                probability_values[index][0] = round(1 - probability_value, 3)
            elif (probability["probability_of"] == f"!{node}" and probability["given"] == f"!{parent}"):
                probability_value = float(probability["equals"])
                probability_values[index][0] = round(probability_value, 3)
                probability_values[index][1] = round(1 - probability_value, 3)
    return BayesianNode(
        label=node,
        probability_values=probability_values,
        parents=parents,
        children=[],
    )


cool-bayesian-networks/bayesian_network.py
from bayesian_node import BayesianNode
import utils

class BayesianNetwork(object):

    def __init__(self, nodes, edges, probabilities):

        self.__nodes = nodes
        self.__edges = edges
        self.__parents = self.__get_parents()
        self.__children = self.__get_children()
        self.__middle_children = list(set(self.__parents) & set(self.__children))

        for node in self.__middle_children:
            self.__parents.remove(node)
            self.__children.remove(node)

        self.__probabilities = self.__process_probabilities(probabilities)
        self.__bayesian_nodes = self.__create_bayesian_nodes()

    def __get_parents(self):
        parents = []
        for node in self.__nodes:
            for edge in self.__edges:
                if (node == edge[0]):
                    parents.append(node)
                    break
        return parents

    def __get_children(self):
        children = []
        for node in self.__nodes:
            for edge in self.__edges:
                if (node == edge[1]):
                    children.append(node)
                    break
        return children

    def __process_probabilities(self, probabilities):

        probabilities_struct = []

        with open(probabilities, "r") as file:
            probabilities = file.readlines()

        for line in probabilities:
            probability_dict = utils.get_probability_dict(line)
            probabilities_struct.append(probability_dict)

        return probabilities_struct

    def __create_bayesian_nodes(self):
        bayesian_nodes = []
        for node in self.__nodes:
            if (node in self.__parents):
                probability_values = [[0, 0]]
                new_node = utils.get_parent_node(
                    node=node,
                    probabilities=self.__probabilities,
                    probability_values=probability_values,
                    children=self.__get_personal_children(node),
                )
                bayesian_nodes.append(new_node)
            # elif (node in self.__children):
            #     parents = self.__get_personal_parents(node)
            #     probability_values = [[0, 0] for _ in range(2 ** len(parents))]
            #     new_node = utils.get_child_node(
            #         node=node,
            #         probabilities=self.__probabilities,
            #         probability_values=probability_values,
            #         parents=parents,
            #     )
            #     bayesian_nodes.append(new_node)
        return bayesian_nodes

    def __get_personal_parents(self, node):
        personal_parents = []
        for edge in self.__edges:
            if (node == edge[1]):
                personal_parents.append(edge[0])
        return personal_parents

    def __get_personal_children(self, node):
        personal_children = []
        for edge in self.__edges:
            if (node == edge[0]):
                personal_children.append(edge[1])
        return personal_children

    def info(self):
        for edge in self.__edges:
            print(f"{edge[0]} -> {edge[1]}")
        for node in self.__bayesian_nodes:
            print(node)


cool-bayesian-networks/bayesianConstruct.py
from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD 
from pgmpy.inference import VariableElimination

class Bayesian:
    def __init__(self):
        self.model = None
        self.infer = None

    # Creación del modelo bayesiano
    def BayesionNetworkConstruct(self,construct):
        self.model = BayesianNetwork(construct)

    # Definición de las probabilidades condicionales
    def AsignValues(self,name,amount,probabilities,evidences = None ,evidences_card = None):
        return TabularCPD(variable=name, variable_card=amount, 
                   values=probabilities,
                   evidence=evidences, evidence_card=evidences_card)
                   
    # Mostrar los factores de la misma
    def showValues(self,cpd):
        print(cpd.values)


    # Adición de las CPDs al modelo
    def AsignToModel(self,values):
        for value in values:
            self.model.add_cpds(value)
    
    # Verificación del modelo
    def checkModel(self):
        check = self.model.check_model()
        if check:
            print("Esta esta completamente descrita")
        else:
            print("Esta no esta completamente descrita")
    
    # Verificación de los nodos
    def checkNodes(self):
        print("Nodes\n",self.model.nodes())

    # Verificación de los edges
    def checkEdges(self):
        print("Edges\n",self.model.edges())

    # mostrar la construccion que tiene
    def showModelConstruct(self,value):
        print("Model\n",self.model.get_cpds(value))        

    # Calcular probabilidad dado elemento
    def calculateProbability(self,variables,evidenced=None):
        self.infer = VariableElimination(self.model)
        query = self.infer.query(variables=variables, evidence=evidenced)
        return query


cool-bayesian-networks/bayesian_construct.py
# Laboratorio 2
# Inteligencia Artificial
# Authors: YongBum Park 20117, Santiago Taracena 20017, Pedro Arriola 20188, Oscar López 20679

from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD
from pgmpy.inference import VariableElimination


class Bayesian(object):
    def __init__(self):
        self.model = None
        self.infer = None

    def bayesian_network_construction(self, construct):
        # Create a Bayesian Network with the specified structure
        self.model = BayesianNetwork(construct)

    def asign_values(
        self, name, amount, probabilities, evidences=None, evidences_card=None
    ):
        # Define the conditional probability distribution for a node
        return TabularCPD(
            variable=name,
            variable_card=amount,
            values=probabilities,
            evidence=evidences,
            evidence_card=evidences_card,
        )

    def show_values(self, cpd):
        # Display the values of a conditional probability distribution
        print(cpd.values)

    def asign_to_model(self, values):
        # Add the conditional probability distributions to the model
        for value in values:
            self.model.add_cpds(value)
        for ucpd in self.model.get_cpds():
            if not self.model.check_model(ucpd):
                print(f"Error al agregar la CPD {ucpd}")

    def check_model(self):
        # Check if the model is valid
        return self.model.check_model()

    def check_nodes(self):
        # Display the nodes in the model
        print("Nodes\n", self.model.nodes())

    def check_edges(self):
        # Display the edges in the model
        print("Edges\n", self.model.edges())

    def show_model_construction(self, value):
        # Display the construction of the specified node in the model
        print("Model\n", self.model.get_cpds(value))

    def calculate_probability(self, variables, evidenced=None):
        # Calculate the probability of the specified variables given the evidence
        self.infer = VariableElimination(self.model)
        query = self.infer.query(variables=variables, evidence=evidenced)
        return query


cool-bayesian-networks/main.py
# Laboratorio 2
# Inteligencia Artificial

# YongBum Park 20117
# Santiago Taracena 20017
# Pedro Arriola 20188
# Oscar López 20679

from bayesian_construct import *

bayesian = Bayesian()

# Network construction
network_structure = [("A", "C"), ("B", "C")]
bayesian.bayesian_network_construction(network_structure)

cpd_a = bayesian.asign_values("A", 2, [[0.3], [0.7]])
cpd_b = bayesian.asign_values("B", 2, [[0.23], [0.77]])
cpd_c = bayesian.asign_values(
    "C", 2, [[0.20, 0.77, 0.10, 0.5], [0.80, 0.23, 0.90, 0.5]], ["A", "B"], [2, 2]
)
# Display conditional probability distribution
bayesian.show_values(cpd_c)

# Add conditional probability distribution to model
bayesian.asign_to_model([cpd_a, cpd_b, cpd_c])

# Validity of model
print(bayesian.check_model())

# Nodes and edges
bayesian.check_edges()
bayesian.check_nodes()

# Model construction for C
bayesian.show_model_construction("C")

# Prob of C given A and B
evidence = {"A": 0, "B": 1}
result = bayesian.calculate_probability(["C"], evidence)
print(result)
